File: huipyo/src/week02/boj1021/Main.java
/**
 *  회전하는 큐
 *  https://www.acmicpc.net/problem/1021
 *  [ 효율성 ]
 *  - 메모리: 17700KB
 *  - 시간 : 212ms
 */

package week02.boj1021;

import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt(), m = scanner.nextInt();
        Queue<Integer> queue = new ArrayDeque<>() { { for (int i = 0; i < m; i++) offer(scanner.nextInt()); } };
        Deque<Integer> deque = new ArrayDeque<>() { { for (int i = 0; i < n; i++) offer(i + 1); } };
        int count = 0;

        while (!queue.isEmpty()) {
            int target = queue.poll();
            int start = new ArrayList<>(deque).indexOf(target);
            int end = deque.size() - start;

            while (!deque.isEmpty() && target != deque.peekFirst()) {
                count++;
                if (start < end) deque.offerLast(deque.pollFirst());
                else deque.offerFirst(deque.pollLast());
            }
            deque.pollFirst();
        }
        System.out.println(count);
        scanner.close();
    }
}

File: huipyo/src/week02/boj1021/Main.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.on('line', (line) => {
    let [n, m] = line.trim().split(' ').map(Number);

    rl.once('line', (queueLine) => {
        var queue = queueLine.trim().split(' ').map(Number);
        var deque = Array.from({ length: n }, (_, i) => i + 1);
        var count = 0;

        while (queue.length > 0) {
            let target = queue.shift();
            let start = deque.indexOf(target);
            let end = deque.length - start;

            while (deque.length > 0 && target !== deque[0]) {
                count++;
                if (start < end) deque.push(deque.shift());
                else deque.unshift(deque.pop());
            }
            deque.shift();
        }
        console.log(count);
        rl.close();
    });
});